draft-irtf-cfrg-frost.md

```diff
@@ -214,22 +214,45 @@ FROST requires the use of a cryptographically secure hash function, generically
 written as H, which functions effectively as a random oracle. For concrete
 recommendations on hash functions which SHOULD be used in practice, see
 {{ciphersuites}}. Using H, we introduce three separate domain-separated hashes,
-H1, H2, and H3, where H1 and H2 map arbitrary byte strings to Scalar elements of
+H1, H2, H3, and H4, where H1, H2, and H4 map arbitrary byte strings to Scalar elements of
 the prime-order group scalar field, and H3 is an alias for H with a domain separator. 
-The details of H1, H2, and H3 vary based on ciphersuite. See {{ciphersuites}}
+The details of H1, H2, H3, and H4 vary based on ciphersuite. See {{ciphersuites}}
 for more details about each.
 
 # Helper functions {#helpers}
 
 Beyond the core dependencies, the protocol in this document depends on the
 following helper operations:
 
+- Nonce generation, {{dep-nonces}}
 - Schnorr signatures, {{dep-schnorr}};
 - Polynomial operations, {{dep-polynomial}};
 - Encoding operations, {{dep-encoding}};
 - Signature binding {{dep-binding-factor}}, group commitment {{dep-group-commit}}, and challenge computation {{dep-sig-challenge}}
 
-This sections describes these operations in more detail.
+These sections describes these operations in more detail.
+
+## Nonce generation {#dep-nonces}
+
+To hedge against a bad RNG, we generate nocnes by sourcing fresh randomness and 
+combine with the secret key, to create a domain-separated hash function from 
+the ciphersuite hash function `H`, `H4`:
+
+~~~
+  nonce_generate(secret):
+  
+  Inputs:
+  - secret, a Scalar in GF(p).
```

Comment:
I don't think this should be a Scalar input -- I think it should be random bytes.

Reply:
I would argue for `Scalar` - no other subroutine takes 'bytes', the closest is a 'message'. The ~closest subroutine to this `nonce_generate` is `compute_challenge()`, which has `Public key, an Element` as an input; similarly here, we expect some `secret key`, which in either singleton or threshold is a `Scalar`.

Reply:
Why does this need to be a secret key as input though? One could, for example, fold in any high-entropy secret established out of band to achieve the same effect, right?

Reply:
A good question!

Reply:
So, this hedge is here against a bad runtime RNG, by hoping/assuming that the RNG used to generate the secret key (share) was possibly better. We give specific requirements/assumptions for how FROST secret shares are generated. Allowing some generic set of bytes vs the secret key has less strict guarantees about the randomness of those bytes. Specifying that this is the secret key (share) gives users less leeway to say, rng the input bytes with the same rng used to source more nonce randomness at runtime.

Reply:
Where do we give requirements for how secret key shares are generated? I thought that was out of scope for this document.

Reply:
![Screenshot_20220521-211017.png](https://user-images.githubusercontent.com/552961/169674108-6b04b9fa-df94-4abc-a366-68ae07340a32.png)

Reply:
That's the requirement for any sampling in the spec, including the nonce, so I don't really think it's relevant to the discussion here? In any case, this is entirely an implementation detail that does not affect interoperability, so whatever we put here cannot possibly be enforced. I think we ought to go with the thing that can be used most broadly, and I currently think making it take in bytes -- which can be the encoded secret key share! -- is the best way to achieve that.

Reply:
It's not just about interop or enforceability but about the first suggestion to an implementor, and this suggestion is not about 'used most broadly', it's about taking learned lessons from ECDSA to hedge against bad RNGs.

Reply:
I'm not really sure we're understanding each other, because the suggestion I've provided would yield the exact same implementation as that which is currently done.

Reply:
no, text that tells a human implementer, pass in your secret key, versus text that says, pass in 32 bytes, does not deterministically result in the same implementation

Reply:
Please take a look at the suggestion I left.

Reply:
I would still enforce this subroutine signature as a Scalar type, not 32 bytes (or equivalent number of Scalar bytes), as it is more likely to result in a function implementation that takes a 'key' type or at least a Scalar type which comes with more checks, and not an array of bytes

Reply:
Although the sourcing of a new random scalar and the immediately turning it into bytes to be hashed could be changed, if this is the only invocation of RandomNonZeroScalar maybe it should be replaced with RandomBytes of something

Reply:
> I would still enforce this subroutine signature as a Scalar type, not 32 bytes (or equivalent number of Scalar bytes), as it is more likely to result in a function implementation that takes a 'key' type or at least a Scalar type which comes with more checks, and not an array of bytes

Well, since this is an implementation detail, we can't force anything to happen really. At best we can nudge folks to do the sensible thing.

In any case, I think we've discussed this through enough. I think this is the wrong way to slice the abstraction for the spec, and I'm willing to be in the rough on that.

Reply:
Yes this is a nudge, cannot enforce but hopefully lead